src/uptane/uptanerepository.cc

```diff
@@ -74,13 +74,29 @@ Json::Value Repository::signVersionManifest(const Json::Value& primary_version_m
 // Check for consistency, signatures are already checked
 bool Repository::verifyMeta(const Uptane::MetaPack& meta) {
   // verify that director and image targets are consistent
-  for (std::vector<Uptane::Target>::const_iterator it = meta.director_targets.targets.begin();
-       it != meta.director_targets.targets.end(); ++it) {
-    std::vector<Uptane::Target>::const_iterator image_target_it;
-    image_target_it = std::find(meta.image_targets.targets.begin(), meta.image_targets.targets.end(), *it);
+  if (meta.director_targets.version() != meta.image_targets.version()) {
+    LOG_ERROR << "Director targets.json version is: " << meta.director_targets.version()
+              << " image targets.json version is: " << meta.image_targets.version();
+    return false;
+  }
+  for (auto director_target : meta.director_targets.targets) {
+    auto image_target_it = std::find_if(meta.image_targets.targets.begin(), meta.image_targets.targets.end(),
+                                        [&director_target](const Uptane::Target& image_target) -> bool {
+                                          if (director_target.length() > image_target.length()) return false;
+                                          if (director_target.filename() != image_target.filename()) return false;
+                                          if (director_target.hashes().size() != image_target.hashes().size())
+                                            return false;
+                                          // TODO: hardware_identifier and release_counter?
+                                          for (auto director_hash : director_target.hashes()) {
+                                            auto found = std::find(image_target.hashes().begin(),
+                                                                   image_target.hashes().end(), director_hash);
+                                            if (found == image_target.hashes().end()) return false;
+                                          }
+                                          return true;
+                                        });
```

Comment:
Can you move this logic out into somewhere that can be isolated and unit tested?  This logic is subtle, and without being able to easily drive it, we are going to have a hard time testing and maintaining it.

A free function that compares a director and images target would be fine, because the comparison is asymmetric.

Reply:
I don't think the free function it is good. We could create Repository in tests, then call verifyMeta, passing MetaPack as argument. 
I will add unit tests for this

Reply:
The free function I was thinking of would just replace the lambda function that had the subtle logic in it. It would be a pure function of the two targets objects (I think?)